persistence/src/main/scala/com/evolutiongaming/akkaeffect/persistence/EventSourcedActorOf.scala

```diff
@@ -0,0 +1,135 @@
+package com.evolutiongaming.akkaeffect.persistence
+
+import akka.actor.Actor
+import akka.persistence.SnapshotSelectionCriteria
+
+import cats.effect.implicits.effectResourceOps
+import cats.effect.{Async, Resource, Ref}
+import cats.syntax.all._
+
+import com.evolutiongaming.akkaeffect._
+import com.evolutiongaming.catshelper.ToFuture
+
+import scala.reflect.ClassTag
+import java.time.Instant
+
+object EventSourcedActorOf {
+
+  /** Describes lifecycle of entity with regards to event sourcing & PersistentActor Lifecycle phases:
+    *
+    *   1. RecoveryStarted: we have id in place and can decide whether we should continue with recovery 
+    *   2. Recovering : reading snapshot and replaying events 
+    *   3. Receiving : receiving commands and potentially storing events & snapshots 
+    *   4. Termination : triggers all release hooks of allocated resources within previous phases
+    *
+    * @tparam S snapshot
+    * @tparam E event
+    * @tparam C command
+    */
+  type Lifecycle[F[_], S, E, C] =
+    Resource[F, RecoveryStarted[F, S, E, Receive[F, Envelope[C], ActorOf.Stop]]]
+
+  def actor[F[_]: Async: ToFuture, S, E, C: ClassTag](
+    eventSourcedOf: EventSourcedOf[F, Lifecycle[F, S, E, C]],
+    persistence: EventSourcedPersistence[F]
+  ): Actor = ActorOf[F] { actorCtx =>
+    for {
+      eventSourced    <- eventSourcedOf(actorCtx).toResource
+      recoveryStarted <- eventSourced.value
+
+      snapshotStore <- persistence.snapshotStore[S](eventSourced).toResource
+      eventStore    <- persistence.eventStore[E](eventSourced).toResource
+
+      snapshot <- snapshotStore.latest.toResource
+
+      recovering <- recoveryStarted(
+        snapshot.map(_.metadata.seqNr).getOrElse(SeqNr.Min),
+        snapshot.map(_.asOffer)
+      )
```

Comment:
I would consider naming the method to something like `#offerSnapshot` or `#recover` instead of `#apply` for better clarity. :thinking:

Reply:
Did I understand correctly that `recoveryStarted` should not be called twice here?

Would it be worth (for better reading / safety?) to use shadowing to prevent this to be done accidentally?

I.e. like this:
```
currentPhase <- eventSourced.value
currentPhase <- currentPhase.offerSnapshot(...)
currentPhase <- currentPhase.completed(...)
```

You will lose the name of the current phase, though, in this case, so there are downsides to this approach too.